File: lib/switch.go
package lib

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"path"

	"github.com/otiai10/copy"
	"github.com/ryanccn/nyoom/config"
	"github.com/ryanccn/nyoom/utils"
)

func Switch(chrome config.Userchrome, profile string) {
	tempDir, err := os.MkdirTemp(os.TempDir(), "nyoom")
	if err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Cloning repository")

	cloneCmd := exec.Command("git", "clone", "--depth=1", chrome.CloneURL, tempDir)
	err = cloneCmd.Run()
	if err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Copying chrome directory")
	newChromeDir := path.Join(profile, "chrome")

	if utils.Exists(newChromeDir) {
		os.RemoveAll(newChromeDir)
	}
	os.Mkdir(newChromeDir, 0755)

	chromeDirectory := path.Join(tempDir, "chrome")
	if !utils.Exists(chromeDirectory) {
		chromeDirectory = tempDir
	}

	err = copy.Copy(chromeDirectory, newChromeDir)
	if err != nil {
		log.Fatalln(err)
	}

	err = os.RemoveAll(tempDir)
	if err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Applying configs to user.js")
	applyUserJs(chrome.Configs, profile)

	fmt.Printf("Switched to %s!\n", chrome.Name)
}

File: src/util/switch.rs
use nanoid::nanoid;
use std::{env, fs, io::Result, path::Path, process::Command};

use crate::config::Userchrome;

fn copy_dir_all(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> Result<()> {
    fs::create_dir_all(&dst)?;

    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let ty = entry.file_type()?;
        if ty.is_dir() {
            copy_dir_all(entry.path(), dst.as_ref().join(entry.file_name()))?;
        } else {
            fs::copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
        }
    }

    Ok(())
}

pub fn switch(userchrome: &Userchrome, profile: String) {
    let temp_path = env::temp_dir().join(nanoid!());

    let mut clone_cmd = Command::new("git");
    clone_cmd.args([
        "clone",
        "--depth=1",
        &userchrome.clone_url,
        temp_path.to_str().unwrap(),
    ]);
    clone_cmd.status().unwrap();

    let new_chrome_dir = Path::new(&profile).join("chrome");

    fs::remove_dir_all(&new_chrome_dir).unwrap();

    let mut cloned_chrome_dir = temp_path.join("chrome");
    if !cloned_chrome_dir.exists() {
        cloned_chrome_dir = temp_path;
    }

    copy_dir_all(&cloned_chrome_dir, &new_chrome_dir).unwrap();
}
